distributed/shuffle/_worker_extension.py

```diff
@@ -250,22 +276,19 @@ async def get_output_partition(self, i: int) -> pd.DataFrame:
         # ^ NOTE: this check isn't necessary, just a nice validation to prevent incorrect
         # data in the case something has gone very wrong
 
-        assert (
-            self.output_partitions_left > 0
-        ), f"No outputs remaining, but requested output partition {i} on {self.local_address}."
         await self.flush_receive()
         try:
-            df = self._read_from_disk(i)
+            data = self._read_from_disk(i)
+            df = convert_partition(data)
             with self.time("cpu"):
                 out = df.to_pandas()
         except KeyError:
             out = self.schema.empty_table().to_pandas()
-        self.output_partitions_left -= 1
         return out
 
-    def _read_from_disk(self, id: int | str) -> pa.Table:
+    def _read_from_disk(self, id: int | str) -> bytes:
         self.raise_if_closed()
-        return self._disk_buffer.read(id)
+        return self._disk_buffer.read(id)[0]
```

Comment:
This is ugly, but it allows us to avoid `pa.Table` in the buffers without rewriting signatures. I would normally clean this up, but since we expect to have to refactor the buffers in the next step, i.e., implementing rechunking based on `p2p`, I would like to avoid duplicated work.

Reply:
maybe

```suggestion
        data = self._disk_buffer.read(id)
        assert len(data) == 1
        return data[0]
```